internal/servers/worker/internal/metric/proxy_ws.go

```diff
@@ -0,0 +1,118 @@
+package metric
+
+import (
+	"bufio"
+	"net"
+	"net/http"
+
+	"github.com/hashicorp/boundary/globals"
+	"github.com/prometheus/client_golang/prometheus"
+)
+
+const (
+	proxyWebsocketSubsystem = "proxy_websocket"
+)
+
+var (
+	msgSizeBuckets = prometheus.ExponentialBuckets(100, 10, 8)
+
+	activeConnCount = prometheus.NewGauge(
+		prometheus.GaugeOpts{
+			Namespace: globals.MetricNamespace,
+			Subsystem: proxyWebsocketSubsystem,
+			Name:      "active_connection_count",
+			Help:      "Count of open websocket connections to Boundary workers.",
+		},
+	)
+
+	bytesReceived = prometheus.NewHistogram(
+		prometheus.HistogramOpts{
+			Namespace: globals.MetricNamespace,
+			Subsystem: proxyWebsocketSubsystem,
+			Name:      "bytes_received",
+			Help:      "Histogram of received bytes for Worker connections.",
+			Buckets:   msgSizeBuckets, // 100 bytes, 1kb, 10kb, 100kb, 1mb, 10mb, 100mb, 1gb
+		},
+	)
+
+	bytesSent = prometheus.NewHistogram(
+		prometheus.HistogramOpts{
+			Namespace: globals.MetricNamespace,
+			Subsystem: proxyWebsocketSubsystem,
+			Name:      "bytes_sent",
+			Help:      "Histogram of sent bytes for Worker connections.",
+			Buckets:   msgSizeBuckets, // 100 bytes, 1kb, 10kb, 100kb, 1mb, 10mb, 100mb, 1gb
+		},
+	)
+)
+
+func MustInitializeProxyMetrics(r prometheus.Registerer) {
+	if r == nil {
+		return
+	}
+	r.MustRegister(activeConnCount, bytesReceived, bytesSent)
+}
+
+func ProxyWsMetricsHandler(next http.Handler) http.Handler {
+	return http.HandlerFunc(func(rw http.ResponseWriter, req *http.Request) {
+		nextRw := rw
+		if rwh, ok := rw.(http.Hijacker); ok {
+			nextRw = &metricsWrapper{ResponseWriter: rw, Hijacker: rwh}
+		}
+		next.ServeHTTP(nextRw, req)
+	})
+}
+
+type wsConn struct {
+	net.Conn
+}
+
+func (wc *wsConn) Read(b []byte) (int, error) {
+	n, err := wc.Conn.Read(b)
+	if err != nil {
+		return n, err
+	}
+	bytesReceived.Observe(float64(len(b))) // The slice is always at least 4096 bytes long (`bufio.ReadWriter` default)
```

Comment:
This is a noteworthy point that should be discussed in this RFC.
The byte slice len will be at least `4096`. It'll consist of the actual data and the rest (if not filled up by the data itself) padded by zeroes.

It's probably worth to discuss if we should remove that padding for our counter as we'll be over-reporting received bytes if we don't.